demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+    eGeneratedRsa,
+    eDeliberatelyInvalid, /* Provisioned using credentials that are meant to trigger an error condition. */
+    eStateUnknown         /* State of the credentials is unknown. */
+} CredentialsProvisioned_t;
+
+
+
+/* GLOBAL VARIABLES */
+CK_SESSION_HANDLE xGlobalSession;
+CK_RV xResult;
+
+CK_OBJECT_HANDLE xPrivateKeyHandle = CK_INVALID_HANDLE;
+CK_OBJECT_HANDLE xPublicKeyHandle = CK_INVALID_HANDLE;
+CK_FUNCTION_LIST_PTR pxGlobalFunctionList;
+
+/*-----------------------------------------------------------*/
+
+/* Declaration of demo function. */
+void vStartKeyProvisioningDemo( void );
+
+/*-----------------------------------------------------------*/
+
+
+/* @brief Random Number Generator Used to Generate CSR
+ *
+ *
+ *	\param[in] pkcs_session     Pointer to PKCS Session Handle
+ *
+ *  \param[in] pucRandom        Starting Point
+ *
+ *  \param[in] xRandomLength    Length
+ *
+ *  \return 0 on success
+ *
+ */
+static int prvRNG( void * pkcs_session,
+                   unsigned char * pucRandom,
+                   size_t xRandomLength )
+{
+    BaseType_t xResult;
+    CK_FUNCTION_LIST_PTR pxP11FunctionList;
+
+    xResult = C_GetFunctionList( &pxP11FunctionList );
+
+    xResult = pxP11FunctionList->C_GenerateRandom( ( *( ( CK_SESSION_HANDLE * ) pkcs_session ) ), pucRandom, xRandomLength );
+
+    if( xResult != 0 )
+    {
+        configPRINTF( ( "ERROR: Failed to generate random bytes %d \r\n", xResult ) );
+        xResult = TLS_ERROR_RNG;
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+/* @brief Alternate Signing Function to be Passed into PK Context Header */
+static int prvPrivateKeySigningCallback( void * pvContext,
+                                         mbedtls_md_type_t xMdAlg,
+                                         const unsigned char * pucHash,
+                                         size_t xHashLen,
+                                         unsigned char * pucSig,
+                                         size_t * pxSigLen,
+                                         int ( * piRng )( void *,
+                                                          unsigned char *,
+                                                          size_t ), /*lint !e955 This parameter is unused. */
+                                         void * pvRng )
+{
+    CK_RV xResult = 0;
+    int lFinalResult = 0;
+    CK_MECHANISM xMech = { 0 };
+    CK_BYTE xToBeSigned[ 256 ];
+    uint8_t ucTemp[ 64 ] = { 0 }; /* A temporary buffer for the pre-formatted signature. */
+    CK_ULONG xToBeSignedLen = sizeof( xToBeSigned );
+
+
+    /* Unreferenced parameters. */
+    ( void ) ( piRng );
+    ( void ) ( pvRng );
+    ( void ) ( xMdAlg );
+
+    /* Sanity check buffer length. */
+    if( xHashLen > sizeof( xToBeSigned ) )
+    {
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+
+    xMech.mechanism = CKM_ECDSA;
```

Comment:
Is this routine ECDSA-only? The defines at the top of this file led me to believe that the demo also supports RSA.

Reply:
Yes, it's ECDSA only. Removed all unnecessary RSA defines. Thanks.